Generalize LTS kernel upgrades to 14.04 and to vagrant

charness opened the pull request:
- Change the logic on what needs upgrading to pick up the LTS HWE kernel on 14.04 and to explicitly install kernel extras only for 13.x.
- Factor out the reboot logic
- Add a `vagrant reload` for vagrant instances, so vagrant-cachier (et al?) works after the reboot

Commit:
Upgrade 14.04 to xenial kernel

Extend the 12.04 -> trusty kernel upgrade to the comparable (hardware
enablement) upgrade of 14.04 -> xenial

linux-image-generic-lts-xenial depends on the corresponding
kernel-extra package, so limit that task to 13.xx distros only.

Commit:
Factor out reboot logic

In preparation for properly supporting vagrant reboots, factor out the
reboot logic into its own file. The logic could be useful elsewhere,
and this also abstracts the concept out of the way of the real job of
upgrading.

Commit:
Use vagrant reload for rebooting a vagrant instance

Rebooting the instance from the machine itself bypasses setup that
`vagrant up' does, such as synchronized folders. This breaks, at
least, package installation with the vagrant-cachier plugin, as its
overlay directory won't be mounted.

Use ansible_ssh_user == 'vagrant' as the signal that a machine is a
vagrant instance.

Commit:
Centralize vagrant detection

- Don't repeat yourself
- Allow explicit setting of `is_vagrant' to override the detection,
  e.g., to support a user other than vagrant on a vagrant instance

Commit:
Trust packages for whether reboot is necessary

Installing a kernel package on Ubuntu creates a
/var/run/reboot-required file. Key off its existence instead of trying
to figure out ourselves whether to reboot.

The `removes' argument to `command' handles conditioning the reboot on
whether the file exists.

For a vagrant instance, fire a `vagrant reload' after rebooting from
on the machine so we still get the setup `vagrant up' does.

dbyron0 commented on tasks/main.yml:
```diff
@@ -11,25 +11,29 @@
 
 # https://docs.docker.com/installation/ubuntulinux/
-- name: Install trusty kernel onto 12.04
+- name: Install backported kernel on pre-16.04 LTS
```
Sorry if I being dense.  What's backported here?

charness replied:
The kernel, from one LTS release to the previous: https://wiki.ubuntu.com/Kernel/LTSEnablementStack

dbyron0 reviewed (commented):
Any chance for automated tests for this?

Commit:
Expand comment on kernel upgrade

- Add to the comment some information on why we're upgrading
- Update the link for docker's redirect and to point to a more
  specific fragment of the doc

dbyron0 commented on tasks/main.yml:
```diff
@@ -10,26 +10,37 @@
+#          corresponding extra package, to address issues with kernels
+#          before 3.19 at the same time.
+#          https://github.com/docker/docker/issues/21704#issuecomment-235365424
+- name: Install HWE kernel on pre-16.04 LTS
```
Thanks much.
